lsp-types/src/Language/LSP/MetaModel/CodeGen.hs

```diff
@@ -0,0 +1,911 @@
+  , fromStringClause       :: TH.Clause
+  , messageDirectionClause :: TH.Clause
+  , messageKindClause      :: TH.Clause
+  }
+
+data NotificationData = NotificationData
+  { methCon                :: TH.Con
+  , singCon                :: TH.Con
+  , paramsEq               :: TH.TySynEqn
+  , registrationOptionsEq  :: TH.TySynEqn
+  , toStringClause         :: TH.Clause
+  , fromStringClause       :: TH.Clause
+  , messageDirectionClause :: TH.Clause
+  , messageKindClause      :: TH.Clause
+  }
+
+data CustomData = CustomData
+  { methCon                :: TH.Con
+  , singCon                :: TH.Con
+  , paramsEq               :: TH.TySynEqn
+  , resultEq               :: TH.TySynEqn
+  , errorDataEq            :: TH.TySynEqn
+  , registrationOptionsEq  :: TH.TySynEqn
+  , toStringClause         :: TH.Clause
+  , fromStringClause       :: TH.Clause
+  , messageDirectionClause :: TH.Clause
+  , messageKindClause      :: TH.Clause
+  }
+
+-- TODO: partial result params
+genMethods :: [Request] -> [Notification] -> CodeGenM [TH.Dec]
+genMethods reqs nots = do
+  mtyN <- lift $ TH.newName "Method"
+  -- These are the type variables for the main type declaration, we will need them for
+  -- some of the constructor declarations
+  fN <- lift $ TH.newName "f"
+  tN <- lift $ TH.newName "t"
+  lift $ addHaddock mtyN "A type representing a LSP method (or class of methods), intended to be used mostly at the type level."
+
+  styN <- lift $ TH.newName "SMethod"
+  lift $ addHaddock styN "A singleton type for 'Method'."
+
+  sstyN <- lift $ TH.newName "SomeMethod"
+  lift $ addHaddock sstyN "A method which isn't statically known."
+  -- We will want to refer to the constructor in some places
+  smcn <- lift $ TH.newName "SomeMethod"
+
+  mpN <- lift $ TH.newName "MessageParams"
+  lift $ addHaddock mpN "Maps a LSP method to its parameter type."
+
+  mrN <- lift $ TH.newName "MessageResult"
+  lift $ addHaddock mrN "Maps a LSP method to its result type."
+
+  edN <- lift $ TH.newName "ErrorData"
+  lift $ addHaddock edN "Maps a LSP method to its error data type."
+
+  roN <- lift $ TH.newName "RegistrationOptions"
+  lift $ addHaddock roN "Maps a LSP method to its registration options type."
+
+  toStringN <- lift $ TH.newName "someMethodToMethodString"
+  lift $ addHaddock toStringN "Turn a 'SomeMethod' into its LSP method string."
+  fromStringN <- lift $ TH.newName "methodStringToSomeMethod"
+  lift $ addHaddock fromStringN "Turn a LSP method string into a 'SomeMethod'."
+
+  mdN <- lift $ TH.newName "messageDirection"
+  lift $ addHaddock mdN "Get a singleton witness for the message direction of a method."
+  mtN <- lift $ TH.newName "messageKind"
+  lift $ addHaddock mtN "Get a singleton witness for the message kind of a method."
+
+  let methodName context fullName =
+        let pieces = T.splitOn "/" fullName
+        -- TODO: don't like this
+        in makeConstrName (Just context) $ foldMap capitalize pieces
```

Comment:
I don't see anything wrong with this? Could probably just default to `Nothing` instead of `Just context` none of the context appears to disambiguous